### PR TITLE
TRD fix digit sorting

### PR DESCRIPTION
**Pull request by bazinski**
- Redo sorting of incoming digits by index. Some digits were assigned to wrong mcm.
- fix linkrecord, remove struct to enable root streaming.
- some other sundry fixes.
sorting by index is needed for removal of timestamp from digits.
It also fixes some wrongly assigned hardware link ID's

### CODE REVIEW

**Comment by bazinski** [comment]
@tdietel can you approve this?

File: DataFormats/Detectors/TRD/include/DataFormatsTRD/LinkRecord.h [hunk 1]
```diff
@@ -23,47 +23,47 @@ namespace o2
...
+        uint16_t stack : 3;
+        uint16_t supermodule : 5;
+      };
+    };
+  };
+
  public:
   LinkRecord() = default;
-  LinkRecord(const uint32_t linkid, int firstentry, int nentries) : mDataRange(firstentry, nentries) { mLinkId.word = linkid; }
+  LinkRecord(const uint32_t linkid, int firstentry, int nentries) : mDataRange(firstentry, nentries) { mLinkId = linkid; }
```
Comment (tdietel) [inline DataFormats/Detectors/TRD/include/DataFormatsTRD/LinkRecord.h:48]: Minor issue: wouldn't it make more sense to have a uint16_t linkid?

File: DataFormats/Detectors/TRD/include/DataFormatsTRD/LinkRecord.h [hunk 2]
```diff
@@ -23,47 +23,47 @@ namespace o2
...
+  const uint32_t getLinkId() { return mLinkId; }
   //TODO come backwith a ccdb lookup.  const uint32_t getLinkHCID() { return mLinkId & 0x7ff; } // the last 11 bits.
-  const uint32_t getSector() { return mLinkId.supermodule; }
-  const uint32_t getStack() { return mLinkId.stack; }
-  const uint32_t getLayer() { return mLinkId.layer; }
-  const uint32_t getSide() { return mLinkId.side; }
+  const uint32_t getSector() { return (mLinkId & 0xf800) >> 11; }
+  const uint32_t getStack() { return (mLinkId & 0x700) >> 8; }
+  const uint32_t getLayer() { return (mLinkId & 0xe0) >> 5; }
+  const uint32_t getSide() { return (mLinkId & 0x10) >> 4; }
```
Comment (tdietel) [inline DataFormats/Detectors/TRD/include/DataFormatsTRD/LinkRecord.h:66]: I'm not sure why these were changed - the old version looked more readable (and safer...) to me.

File: DataFormats/Detectors/TRD/include/DataFormatsTRD/LinkRecord.h [hunk 3]
```diff
@@ -72,9 +72,9 @@ class LinkRecord
   void printStream(std::ostream& stream);
 
  private:
-  LinkId mLinkId;
+  uint16_t mLinkId;
```
Comment (tdietel) [inline DataFormats/Detectors/TRD/include/DataFormatsTRD/LinkRecord.h:75]: Maybe I don't understand the intention here. This looks like you change the LinkRecord to use the uint16_t directly, but lines 33-44 still have the union/struct combo. Maybe I need to look at the file as a whole...

File: DataFormats/Detectors/TRD/src/DataFormatsTRDLinkDef.h [hunk 4]
```diff
@@ -16,6 +16,7 @@
 
 #pragma link C++ class o2::trd::TriggerRecord + ;
 #pragma link C++ class o2::trd::LinkRecord + ;
+#pragma link C++ struct o2::trd::LinkRecord::LinkId + ;
```
Comment (tdietel) [inline DataFormats/Detectors/TRD/src/DataFormatsTRDLinkDef.h:19]: Is this struct actually used?

File: Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx [hunk 5]
```diff
@@ -661,7 +597,8 @@ void TRDDPLTrapSimulatorTask::run(o2::framework::ProcessingContext& pc)
     LOG(info) << "Raw data words written = " << mTotalRawWordsWritten << " with a vector size = " << rawdata.size();
   }
   LOG(debug) << "END OF RUN .............";
-
+  //TODO does anyone care to have the digits to tracklet mapping. Do we then presort the digits inline with the index or send both digits and sorted index.
```
Comment (tdietel) [inline Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx:600]: Good question. One could get very close by just knowing the MCM and the tracklet position - if more is needed, we can get it from the debug stream.

**Review by tdietel — APPROVED** [review]
I think the LinkId stuct within the LinkRecord class is not necessary. Otherwise things look good.

File: DataFormats/Detectors/TRD/src/DataFormatsTRDLinkDef.h (hunk 4 above)
Comment (bazinski) [reply DataFormats/Detectors/TRD/src/DataFormatsTRDLinkDef.h:19]: oops not anymore

File: DataFormats/Detectors/TRD/include/DataFormatsTRD/LinkRecord.h (hunk 3 above)
Comment (bazinski) [reply DataFormats/Detectors/TRD/include/DataFormatsTRD/LinkRecord.h:75]: something weird when rebasing the struct is supposed to be gone.

File: DataFormats/Detectors/TRD/include/DataFormatsTRD/LinkRecord.h (hunk 1 above)
Comment (bazinski) [reply DataFormats/Detectors/TRD/include/DataFormatsTRD/LinkRecord.h:48]: yes.

File: DataFormats/Detectors/TRD/include/DataFormatsTRD/LinkRecord.h (hunk 2 above)
Comment (bazinski) [reply DataFormats/Detectors/TRD/include/DataFormatsTRD/LinkRecord.h:66]: I agree, but with the struct root could not serialise it, let me try one last thing.

**Pull request merged**